### ISSUE TITLE
[expo-updates] Attempt to invoke virtual method 'void expo.modules.updates.db.entity.AssetEntity.setLaunchAsset(boolean)' on a null object reference

### ISSUE THREAD

**Issue by thespacemanatee** [expo/expo#28168]
### Minimal reproducible example

Difficult to reproduce this with a fresh project since it needs to be installed by a large user base

### What platform(s) does this occur on?

Android

### Did you reproduce this issue in a development build?

No (only happens in a standalone build)

### Summary

Ever since we migrated to `expo-updates` a small number of our users are experiencing the following error:

![image](https://github.com/expo/expo/assets/6837599/d1025c21-1035-46a9-bd65-688a1f6f753b)

These are the devices that it affects:

![image](https://github.com/expo/expo/assets/6837599/37986f66-c124-4e9e-8b5b-27ba40142a46)

We don't see any Google devices affected by this bug so perhaps its vendor-specific.

The bug occurs here in `expo/modules/updates/db/dao/UpdateDao.kt:67`:

<img width="385" alt="Screenshot 2024-04-12 at 11 35 00 AM" src="https://github.com/expo/expo/assets/6837599/10fd877d-2aab-4e3d-9921-28bd975dcbb4">

`assetEntity` should not be `null` but for some reason the Room query here returns `null`.

### Environment

```sh
expo-env-info 1.2.0 environment info:
  System:
    OS: macOS 14.4
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.18.0 - ~/.asdf/installs/nodejs/18.18.0/bin/node
    Yarn: 1.22.21 - ~/.asdf/installs/nodejs/18.18.0/bin/yarn
    npm: 9.8.1 - ~/.asdf/plugins/nodejs/shims/npm
    Watchman: 2024.01.22.00 - /opt/homebrew/bin/watchman
  Managers:
    CocoaPods: 1.14.3 - /Users/honeycomb/.asdf/shims/pod
  SDKs:
    iOS SDK:
      Platforms: DriverKit 23.4, iOS 17.4, macOS 14.4, tvOS 17.4, visionOS 1.1, watchOS 10.4
  IDEs:
    Xcode: 15.3/15E204a - /usr/bin/xcodebuild
  npmGlobalPackages:
    eas-cli: 7.8.0
  Expo Workflow: bare
```

### Expo Doctor Diagnostics

```sh
Detailed check results:

Expected package @expo/config-plugins@~7.8.0
Found invalid:
  @expo/config-plugins@6.0.2
  (for more info, run: npm why @expo/config-plugins)
Advice: Upgrade dependencies that are using the invalid package versions0.

The package "expo-dev-menu" should not be installed directly in your project. It is a dependency of other Expo packages, which will install it automatically as needed.

The following packages should be updated for best compatibility with the installed expo version:
  @react-native-community/datetimepicker@7.6.2 - expected version: 7.6.1
  @sentry/react-native@5.19.3 - expected version: 5.19.1
  @shopify/flash-list@1.6.4 - expected version: 1.6.3
  expo@50.0.12 - expected version: ~50.0.15
  expo-camera@14.0.6 - expected version: ~14.1.2
  expo-dev-client@3.3.9 - expected version: ~3.3.11
  expo-updates@0.24.11 - expected version: ~0.24.12
  lottie-react-native@6.6.0 - expected version: 6.5.1
  react-native-gesture-handler@2.15.0 - expected version: ~2.14.0
  react-native-get-random-values@1.10.0 - expected version: ~1.8.0
  react-native-reanimated@3.8.1 - expected version: ~3.6.2
  jest-expo@50.0.3 - expected version: ~50.0.4
Your project may not work correctly until you install the correct versions of the packages.
Found outdated dependencies
Advice: Use 'npx expo install --check' to review and upgrade your dependencies.

One or more checks failed, indicating possible issues with the project.
```

**Comment by CosmicodeStudio** [comment]
We're seeing the same issue on our app, did you find any solution?

**Comment by thespacemanatee** [comment]
> We're seeing the same issue on our app, did you find any solution?

Not yet, waiting for a response from @douglowder. We're paying quite a bit for updaters so it would be nice for this to be looked into!

**Comment by douglowder** [comment]
@thespacemanatee thanks for flagging the issue, I will pick this up today.

**Comment by douglowder** [comment]
@thespacemanatee we are looking at this issue internally.

The linked PR #28264 is intended to prevent a crash from occurring, and we believe it is safe. However, it will not prevent the bug from occurring, and we would like to do some more research, so we can understand exactly what is causing a null asset to be returned from the DB (this is something that should never happen).

In the meantime, you are welcome to try out the change in that PR, with the understanding that it is experimental and not going to be merged to production.

**Comment by thespacemanatee** [comment]
@douglowder thanks for looking into the problem. I'd like to add some information that might help. We are seeing these crashes for users that are on our latest builds that we do not have OTA updates for. Hence, I'm assuming that these crashes are happening with the embedded update. This sounds problematic if true since they are probably unable to use the app. I've yet to verify if this crash will happen for the same user consistently. Hope this helps in some way!

**Comment by elliotmjackson** [comment]
@douglowder I want to provide some impact context here. Im seeing a 10-20% degradation in new user sign up as a result of this bug as I rollout a new version of our app. If theres a latest-stable version of expo-updates I would greatly appreciate this information

**Comment by douglowder** [comment]
The latest stable version of `expo-updates` is 0.24.12, which should be run with SDK 50 (`expo@50.0.17`).

**Comment by wschurman** [comment]
@thespacemanatee - were you able to find an install of the app that is crashing consistently? Or an idea of how I might repro? 

@elliotmjackson - when you say "as I rollout a new version of our app" does that mean you're seeing it for users that are receiving an update? Do you know if the users are on an embedded update or if they already have downloaded one or more updates before seeing this crash?

Unfortunately, without a repro, all that can really be done is guess and treat the symptoms. The PR that @douglowder put up does treat the symptoms and fix the crash in the sense that the app will no longer crash, but it will cause the "emergency launch" mechanism to kick in, and while this may be the behavior we want, it would be better if we could isolate the crash so that one doesn't see a large number of emergency launches.

**Comment by wschurman** [comment]
@elliotmjackson - just double checking, it's only happening for a subset of your users? Is it happening for all Android users or only a subset of them?

**Comment by elliotmjackson** [comment]
Approximately 20% of our users are experiencing issues. Initially, everything appears normal, but the situation deteriorates rapidly as the crash rate climbs. You can observe from the data that the daily signup success rate begins on par with our standard app (indicated in yellow), but then it starts to decline. I've pinpointed that this degradation occurs as soon as the app is launched.
![Screenshot 2024-04-30 at 4 17 02 PM](https://github.com/expo/expo/assets/13633636/487a36f8-c49d-49dd-8e0a-1b8e8fe17131)

**Comment by wschurman** [comment]
@elliotmjackson - what percentage of your users are android vs iOS? or, asking another way, is the 20% all your android users or a subset of your android users?

**Comment by elliotmjackson** [comment]
I have not published the iOS app yet. The vast majority of our users are on android at ~90%

**Comment by wschurman** [comment]
Reopening for future investigation if possible.

**Comment by wschurman** [comment]
Cherry-picked fix (ish) to 50 as well.

**Comment by elliotmjackson** [comment]
@wschurman I wanted to provide you with an update. Four days ago, I implemented the changes from PR #28264 in my app and rolled them out to about 10% of my users. This updated version is running alongside a version of the app that does not use expo-updates. Previously, we observed a significant drop in new user sign-ups. I can now confirm that this issue no longer appears after the recent update. While I understand that this doesn't address the root cause of the problem you're facing, please be assured that it's no longer impacting my end users. Thats good enough for me, thanks for your support!
